velox/dwio/common/tests/TestParallelBufferedInput.cpp

```diff
@@ -0,0 +1,190 @@
+
+  ASSERT_EQ(0, ioStats->rawOverreadBytes());
+}
+
+class TestReadFile : public ReadFile {
+ public:
+  explicit TestReadFile(std::string_view file) : file_(file) {}
+  std::string_view pread(uint64_t offset, uint64_t length, void* buf)
+      const override {
+    // Assume read latency is 1 millisecond per MB.
+    int64_t latency = length >> 20;
+    std::this_thread::sleep_for(std::chrono::milliseconds(latency));
+    return {static_cast<char*>(buf), length};
+  }
+  bool shouldCoalesce() const override {
+    return false;
+  }
+  uint64_t size() const override {
+    return 1 << 30;
+  }
+  uint64_t memoryUsage() const override {
+    return 0;
+  }
+  std::string getName() const override {
+    return "dummy";
+  }
+  uint64_t getNaturalReadSize() const override {
+    return 1 << 20;
+  }
+
+ private:
+  std::string_view file_;
+};
+
+TEST(TestParallelBufferedInput, simpleBenchmark) {
+  std::string filename = "test-file";
+  const int32_t kOneMB = 1 * 1024 * 1024;
+  const int32_t threads = 4;
+  int32_t loadQuantum = 8 * kOneMB;
+  auto executor = std::make_unique<folly::IOThreadPoolExecutor>(threads);
+  auto pool = memory::getDefaultMemoryPool();
+  auto readFile = std::make_shared<TestReadFile>(filename);
+
+  std::vector<Region> regions;
+
+  // 10 regions, each has 2MB length and 2MB distance, eg [0, 2MB], [4MB,
+  // 2MB]...
+  for (uint64_t iter = 0, cursor = 0, length = 2 * kOneMB; iter < 10;
+       ++iter, cursor += 4 * kOneMB) {
+    regions.emplace_back(cursor, length);
+  }
+  // 20 regions, each has 14MB length and 2MB distance, eg [0, 14MB], [16MB,
+  // 14MB]...
+  for (uint64_t iter = 0, cursor = 0, length = 14 * kOneMB; iter < 20;
+       ++iter, cursor += 16 * kOneMB) {
+    regions.emplace_back(cursor, length);
+  }
+
+  auto stats = std::make_shared<IoStatistics>();
+  auto parallelStats = std::make_shared<IoStatistics>();
+  auto parallelFunc = [&]() {
+    ParallelBufferedInput parallelInput(
+        readFile,
+        *pool,
+        MetricsLog::voidLog(),
+        parallelStats,
+        executor.get(),
+        loadQuantum);
+    for (const auto& region : regions) {
+      parallelInput.enqueue({region.offset, region.length});
+    }
+    parallelInput.load(LogType::TEST);
+  };
+  auto normalFunc = [&]() {
+    BufferedInput input(readFile, *pool, MetricsLog::voidLog(), stats.get());
+    for (const auto& region : regions) {
+      input.enqueue({region.offset, region.length});
+    }
+    input.load(LogType::TEST);
+  };
+
+  // Execute 4 times test in alternating order, compare ioStats.
+  bool parallelFist = false;
+  for (int32_t iter = 0; iter < 4; iter++) {
+    if (iter % 2 == 0) {
+      parallelFist = true;
+    } else {
+      parallelFist = false;
+    }
+    if (parallelFist) {
+      parallelFunc();
+      normalFunc();
+    } else {
+      normalFunc();
+      parallelFunc();
+    }
+    ASSERT_EQ(stats->rawBytesRead(), parallelStats->rawBytesRead());
```

Comment:
let's also test with actual expected values here since they are deterministic. If IoStatistics has a bug, then we might be comparing wrong values and still pass the test.

Reply:
We already have related test.
Above test `parallelLoad` focus on verifying data correctness and IoStatistics, this test focus on performance.